libraries/plugins/private_message/private_message_plugin.cpp

```diff
@@ -0,0 +1,185 @@
+                while (itr != idx.end() && tmp.limit && itr->to == tmp.to) {
+                    result.inbox.push_back(*itr);
+                    ++itr;
+                    --tmp.limit;
+                }
+
+                return result;
+            }
+
+            outbox_r private_message_plugin::private_message_plugin_impl::get_outbox(const outbox_a &out) const {
+                FC_ASSERT(out.limit <= 100);
+                outbox_a tmp = out;
+                outbox_r result;
+                const auto &idx = _db.get_index<message_index>().indices().get<by_from_date>();
+
+                auto itr = idx.lower_bound(std::make_tuple(tmp.from, tmp.newest));
+                while (itr != idx.end() && tmp.limit && itr->from == tmp.from) {
+                    result.outbox.push_back(*itr);
+                    ++itr;
+                    --tmp.limit;
+                }
+                return result;
+            }
+
+            private_message_plugin::private_message_plugin_impl::~private_message_plugin_impl() {
+                return;
+            }
+
+
+            void private_message_evaluator::do_apply(const private_message_operation &pm) {
+                database &d = db();
+
+                const flat_map <std::string, std::string> &tracked_accounts = _plugin->tracked_accounts();
+
+                auto to_itr = tracked_accounts.lower_bound(pm.to);
+                auto from_itr = tracked_accounts.lower_bound(pm.from);
+
+                FC_ASSERT(pm.from != pm.to);
+                FC_ASSERT(pm.from_memo_key != pm.to_memo_key);
+                FC_ASSERT(pm.sent_time != 0);
+                FC_ASSERT(pm.encrypted_message.size() >= 32);
+
+                if (!tracked_accounts.size() ||
+                    (to_itr != tracked_accounts.end() && pm.to >= to_itr->first &&
+                     pm.to <= to_itr->second) ||
+                    (from_itr != tracked_accounts.end() &&
+                     pm.from >= from_itr->first && pm.from <= from_itr->second)) {
+                    d.create<message_object>([&](message_object &pmo) {
+                        pmo.from = pm.from;
+                        pmo.to = pm.to;
+                        pmo.from_memo_key = pm.from_memo_key;
+                        pmo.to_memo_key = pm.to_memo_key;
+                        pmo.checksum = pm.checksum;
+                        pmo.sent_time = pm.sent_time;
+                        pmo.receive_time = d.head_block_time();
+                        pmo.encrypted_message.resize(pm.encrypted_message.size());
+                        std::copy(pm.encrypted_message.begin(), pm.encrypted_message.end(),
+                                  pmo.encrypted_message.begin());
+                    });
+                }
+            }
+
+            private_message_plugin::private_message_plugin()
+                    : my(new private_message_plugin::private_message_plugin_impl(*this)) {
```

Comment:
Plugin  has its custom life cycle . Not use RAII.

Reply:
In the code snippet above we use raii technique for impl. Impl does not exist out of plugin class. I do not quite understand the issue.

Reply:
Got it. Fixed.